subprojects/execution/src/main/java/org/gradle/internal/execution/impl/steps/CacheStep.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.impl.steps;
+
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.ImmutableSortedSet;
+import org.gradle.caching.BuildCacheKey;
+import org.gradle.caching.internal.command.BuildCacheCommandFactory;
+import org.gradle.caching.internal.command.BuildCacheLoadListener;
+import org.gradle.caching.internal.controller.BuildCacheController;
+import org.gradle.caching.internal.origin.OriginMetadata;
+import org.gradle.caching.internal.packaging.CacheableTree;
+import org.gradle.caching.internal.packaging.UnrecoverableUnpackingException;
+import org.gradle.internal.execution.ExecutionOutcome;
+import org.gradle.internal.execution.ExecutionResult;
+import org.gradle.internal.execution.OutputChangeListener;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+
+public class CacheStep<C extends CachingContext> implements Step<C> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CacheStep.class);
+
+    private final BuildCacheController buildCache;
+    private final OutputChangeListener outputChangeListener;
+    private final BuildCacheCommandFactory commandFactory;
+    private final Step<? super C> delegate;
+
+    public CacheStep(
+            BuildCacheController buildCache,
+            OutputChangeListener outputChangeListener,
+            BuildCacheCommandFactory commandFactory,
+            Step<? super C> delegate
+    ) {
+        this.buildCache = buildCache;
+        this.outputChangeListener = outputChangeListener;
+        this.commandFactory = commandFactory;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public org.gradle.internal.execution.ExecutionResult execute(C context) {
+        return context.getCacheHandler()
+            .load(cacheKey -> load(context.getWork(), cacheKey))
+            .map(loadResult -> (ExecutionResult) new ExecutionResult() {
```

Comment:
I think this class would deserve an actual type, something like `LoadedFromCache`. Then this line would simplify to `.map(LoadedFromCache::new)`, right?

Reply:
I think you'd still need the cast because of the `orElseGet()`. Let's do this as part of reviewing the codebase for Java 8 compliance later.